[dynamo] add sourceless builder for `types.MethodType`

XuehaiPan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #148151
* #113258
* #113257
* __->__ #147880



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

XuehaiPan committed:
Update

[ghstack-poisoned]

jansel reviewed (changes requested):
Is this safe without a source?  Don't we need a guard?

This also needs a test.

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
> Is this safe without a source? Don't we need a guard?

I updated the implementation. It should be properly guarded.

> This also needs a test.

Added with comments.

XuehaiPan commented:
@pytorchbot merge

wdvr commented:
@pytorchmergebot revert -m 'failing trunk tests' -c ghfirst

test_sparse_semi_structured.py::SparseSemiStructuredTensorCompileTest::test_sp24_compile [GH job link](https://github.com/pytorch/pytorch/actions/runs/13547778883/job/37868524507) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/08f4c1a2332921e57c782c80a66b2adc9cdc0575)

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
@pytorchbot merge

XuehaiPan commented:
@pytorchbot merge

XuehaiPan commented:
@pytorchbot merge -f "long macOS ci queue"

XuehaiPan commented:
@pytorchbot merge -f "long macOS ci queue"
